cmd/rehearse/main.go

```diff
@@ -0,0 +1,222 @@
+	"k8s.io/test-infra/prow/config"
+	prowflagutil "k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pjutil"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func getJobsToExecute(ca *config.Agent) config.JobConfig {
+	var youHaveOneJob config.Presubmit
+	for _, job := range ca.Config().Presubmits["openshift/ci-operator"] {
+		if job.Name == "pull-ci-openshift-ci-operator-master-build" {
+			youHaveOneJob = job
+			break
+		}
+	}
+
+	jobs := config.JobConfig{
+		Presubmits: map[string][]config.Presubmit{"openshift/ci-operator": []config.Presubmit{youHaveOneJob}},
+	}
+	return jobs
+}
+
+func makeRehearsalPresubmit(source *config.Presubmit, repo string, prNumber int) (*config.Presubmit, error) {
+	var rehearsal config.Presubmit
+	deepcopy.Copy(&rehearsal, source)
+
+	rehearsal.Name = fmt.Sprintf("rehearse-%d-%s", prNumber, source.Name)
+	rehearsal.Context = fmt.Sprintf("ci/rehearse/%s/%s", repo, strings.TrimPrefix(source.Context, "ci/prow/"))
+
+	if len(source.Spec.Containers) != 1 {
+		return nil, fmt.Errorf("Cannot rehearse jobs with more than 1 container in Spec")
+	}
+	container := source.Spec.Containers[0]
+
+	if len(container.Command) != 1 || container.Command[0] != "ci-operator" {
+		return nil, fmt.Errorf("Cannot rehearse jobs that have Command different from simple 'ci-operator'")
+	}
+
+	for _, arg := range container.Args {
+		if strings.HasPrefix(arg, "--git-ref") || strings.HasPrefix(arg, "-git-ref") {
+			return nil, fmt.Errorf("Cannot rehearse jobs that call ci-operator with '--git-ref' arg")
+		}
+	}
+
+	if len(source.Branches) != 1 {
+		return nil, fmt.Errorf("Cannot rehearse jobs that run over multiple branches")
+	}
+	branch := strings.TrimPrefix(strings.TrimSuffix(source.Branches[0], "$"), "^")
+
+	gitrefArg := fmt.Sprintf("--git-ref=%s@%s", repo, branch)
+	rehearsal.Spec.Containers[0].Args = append(source.Spec.Containers[0].Args, gitrefArg)
+
+	return &rehearsal, nil
+}
+
+func loadClusterConfig() (*rest.Config, error) {
+	clusterConfig, err := rest.InClusterConfig()
+	if err == nil {
+		return clusterConfig, nil
+	}
+
+	credentials, err := clientcmd.NewDefaultClientConfigLoadingRules().Load()
+	if err != nil {
+		return nil, fmt.Errorf("could not load credentials from config: %v", err)
+	}
+
+	clusterConfig, err = clientcmd.NewDefaultClientConfig(*credentials, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not load client configuration: %v", err)
+	}
+	return clusterConfig, nil
+}
+
+func submitRehearsal(job *config.Presubmit, pr *github.PullRequest, namespace string, logger *logrus.Entry) (*pjapi.ProwJob, error) {
+	pj := pjutil.NewPresubmit(*pr, "", *job, "")
+	logger.WithFields(pjutil.ProwJobFields(&pj)).Info("Submitting a new prowjob.")
+
+	config, err := loadClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not load cluster config")
+	}
+
+	cs, err := pjclientset.NewForConfig(config)
```

Comment:
Should be done once at startup

Reply:
Yeah it's in the WIP items. I was in a hurry yesterday and wanted to get this out.